server/internal/cache/blob/digest.go

```diff
@@ -0,0 +1,97 @@
+package blob
+
+import (
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"slices"
+	"strings"
+)
+
+var (
+	ErrInvalidDigest = errors.New("invalid digest")
+)
+
+// Digest is a blob identifier that is the SHA-256 hash of a blob's content.
+//
+// It is comparable and can be used as a map key.
+type Digest struct {
+	sum [32]byte
+}
+
+// ParseDigest parses a digest from a string. If the string is not a valid
+// digest, a call to the returned digest's IsValid method will return false.
+//
+// The input string may be in one of two forms:
+//
+//   - ("sha256-<hex>"), where <hex> is a 64-character hexadecimal string.
+//   - ("sha256:<hex>"), where <hex> is a 64-character hexadecimal string.
+//
+// The [Digest.String] method will return the canonical form of the
+// digest, "sha256:<hex>".
+func ParseDigest[S ~[]byte | ~string](v S) (Digest, error) {
+	s := string(v)
+	i := strings.IndexAny(s, ":-")
+	var zero Digest
+	if i < 0 {
+		return zero, ErrInvalidDigest
+	}
+
+	prefix, sum := s[:i], s[i+1:]
+	if prefix != "sha256" || len(sum) != 64 {
+		return zero, ErrInvalidDigest
+	}
+
+	var d Digest
+	_, err := hex.Decode(d.sum[:], []byte(sum))
+	if err != nil {
+		return zero, ErrInvalidDigest
+	}
+	return d, nil
+}
+
+func DigestFromBytes[S ~[]byte | ~string](v S) Digest {
```

Comment:
Does the generic make sense here given the function name? vs 
```suggestion
func DigestFromBytes(b []byte) Digest {
```

Reply:
Strings are bytes too.

Reply:
Is this pattern common (or maybe new?) I feel like in the Go standard library everywhere I see functions with the `Bytes` suffix it's explicitly to handle `[]byte` but maybe that's changed since generics? I don't think we're using the `string` version of this anyways.. marking as unresolved ICYMI

Reply:
As mentioned before here https://github.com/ollama/ollama/pull/9294#pullrequestreview-2638347834

I aim only to copy the code over largely unmodified. As PRs progress we can nit at naming.

Reply:
https://github.com/ollama/ollama/blob/1628b398372d8928846612b6e6a1fddd349ea7b7/server/internal/client/ollama/registry_test.go#L116